doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -1766,6 +1820,111 @@ The following errors apply to both string and numeral notations:
 
      .. todo note on "single qualified identifiers" https://github.com/coq/coq/pull/11718#discussion_r415076703
 
+.. example:: Number Notation for a non inductive type
+
+   The following example parses and prints :g:`prod unit (prod unit unit)` as :g:`3`.
+   The inductive :g:`I` will be used as :n:`@qualid__ind`.
+
+   .. coqtop:: in
+
+      Inductive I := Iempty : I | Iunit : I | Iprod : I -> I -> I.
+
+   We then define :n:`@qualid__parse` and :n:`@qualid__print`
+
+   .. coqtop:: in
+
+      Definition of_uint (x : Number.uint) : I :=
+        let fix f n := match n with
+          | O => Iempty | S O => Iunit
+          | S n => Iprod Iunit (f n) end in
+        f (Nat.of_num_uint x).
+
+      Definition to_uint (x : I) : Number.uint :=
+        let fix f i := match i with
+          | Iempty => O | Iunit => 1
+          | Iprod i1 i2 => f i1 + f i2 end in
+        Nat.to_num_uint (f x).
+
+      Inductive prod (A : Set) (B : Set) : Set := pair : A -> B -> prod A B.
+
+   the number notation itself
+
+   .. coqtop:: in
+
+      Notation nSet := Set (only parsing).
+      Number Notation nSet via I using (Empty_set Iempty, unit Iunit, prod Iprod)
+        of_uint to_uint : type_scope.
+
+   and check the printer
+
+   .. coqtop:: all
+
+      Local Open Scope type_scope.
+      Check prod unit (prod unit unit).
+
+   and the parser
+
+   .. coqtop:: all
+
+      Set Printing All.
+      Check 3.
+
+.. example:: Number Notation with implicit arguments
```

Comment:
How about informally describing the number notation used here?

Reply:
done